### PR TITLE
add mxnet

### PR DESCRIPTION
**Pull request by ngam**
- Skipping all cuda builds for now
- Relying on upstream (not conda-forge) tvm (libtvm, etc.)
- Will split into lib and python pkg in feedstock
- Considering relaxing the deps/pinning
- Need to add more runtime deps explicitly (libprotobuf, libopencv, etc.)

---

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by ngam** [comment]
Sorry @hmaarrfk, I messed up the other PR. This is the new location. I also added some patches to avoid rebuilding already-existing libraries, though I am not sure how well they will work. The OneDNN patch worked well, I am not sure about the last two (especially tvm)

**Comment by hmaarrfk** [comment]
we are simply not allowing depending on external channels. please remove the extra channel from the yaml file.

**Comment by ngam** [comment]
> we are simply not allowing depending on external channels. please remove the extra channel from the yaml file.

which channel? That's for the RC release.

**Comment by hmaarrfk** [comment]
You simply cannot have 

```
channel_targets:
   - conda-forge mxnet_rc
```

**Comment by ngam** [comment]
> You simply cannot have
> 
> ```
> channel_targets:
>    - conda-forge mxnet_rc
> ```

More than happy to remove it; but how can I have this go to prelease channel? I followed this: https://github.com/conda-forge/matplotlib-feedstock/blob/rc/recipe/conda_build_config.yaml and this https://conda-forge.org/docs/maintainer/knowledge_base.html#pre-release-builds

**Comment by hmaarrfk** [comment]
I wouldn't worry about having a pre-release channel.

For linux, I strongly recommend you manage your own pre-releases once you get the base feedstock made.

matplotlib is a core package of the python ecosystem, and much discussion has been had around releasing it correctly.

I don't think you want this burden for mxnet. Best to keep it simple.

**Comment by ngam** [comment]
Okay! Sounds good to me!

**Comment by ngam** [comment]
You never responded firmly: Would you like to maintain this with me?

**Comment by hmaarrfk** [comment]
I unfortunately do not have any incentives to maintain this ATM. Happy to help where I can through suggestions, but I don't want to maintain.

File: recipes/mxnet/conda_build_config.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+MACOSX_DEPLOYMENT_TARGET:  # [osx]
+  - '10.13'  # [osx]
+cudnn:
+  - '8'
+
+channel_targets:
+  - conda-forge mxnet_rc
```
Comment (ngam) [inline recipes/mxnet/conda_build_config.yaml:7]:
```suggestion

```

as per suggestion in comment

File: recipes/mxnet/build.sh [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+cmake \
+       -DCMAKE_PREFIX_PATH=$PREFIX \
+       -DCMAKE_BUILD_TYPE=Release \
+       -DCMAKE_INSTALL_LIBDIR="lib" \
+       -DCMAKE_INSTALL_PREFIX=$PREFIX \
+       -DCMAKE_INCLUDE_PATH=$PREFIX/include \
+       -DUSE_CPP_PACKAGE=ON \
+       -DUSE_OPENMP=ON \
+       -DBLA_STATIC=OFF \
```
Comment (ngam) [inline recipes/mxnet/build.sh:35]:
```suggestion
       -DBLA_STATIC=OFF \
       -DBUILD_SHARED_LIBS=ON \
```

add -DBUILD_SHARED_LIBS=ON

File: recipes/mxnet/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    # for library pkg, libmxnet
+    - openblas
+    - llvm-openmp  # [unix]
+    - mkl  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
+    - mkl-include  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
+    - libopencv
+    - graphviz
+    - onednn
+    - cudnn  # [cuda_compiler_version != "None"]
+    - nccl  # [cuda_compiler_version != "None"]
```
Comment (ngam) [inline recipes/mxnet/meta.yaml:94]:
```suggestion
    - nccl  # [cuda_compiler_version != "None"]
    - cutensor  # [cuda_compiler_version != "None"]
```

File: recipes/mxnet/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  run:
+    # for library pkg, libmxnet 
+    - {{ pin_compatible('openblas') }}
+    - {{ pin_compatible('llvm-openmp') }}  # [unix]
+    - {{ pin_compatible('mkl') }}  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
+    - {{ pin_compatible('libopencv') }}
+    - {{ pin_compatible('graphviz') }}
+    - {{ pin_compatible('onednn') }}
+    - {{ pin_compatible('cudnn') }}  # [cuda_compiler_version != "None"]
+    - {{ pin_compatible('nccl') }}  # [cuda_compiler_version != "None"]
```
Comment (ngam) [inline recipes/mxnet/meta.yaml:118]:
```suggestion
    - {{ pin_compatible('nccl') }}  # [cuda_compiler_version != "None"]
    - {{ pin_compatible('cutensor') }}  # [cuda_compiler_version != "None"]
```

File: recipes/mxnet/build.sh [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+   else
+       cp config/distribution/darwin_cpu_mkl.cmake config.cmake
+       export MKL_INCLUDE_DIR=${CONDA_PREFIX}/include
+   fi
+elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+       # if want to add mkl to cuda builds:
+       # CMAKE_ARGS_BLAS=-DUSE_BLAS='mkl'
+       # export MKL_INCLUDE_DIR=${CONDA_PREFIX}/include
+   if [[ "$cuda_compiler_version" == "None" ]]; then
+       export MKL_INCLUDE_DIR=${CONDA_PREFIX}/include
```
Comment (ngam) [inline recipes/mxnet/build.sh:16]:
```suggestion
       CMAKE_ARGS_BLAS=-DUSE_BLAS='mkl'
       export MKL_INCLUDE_DIR=${CONDA_PREFIX}/include
   if [[ "$cuda_compiler_version" == "None" ]]; then
```

File: recipes/mxnet/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    - ccache
+    - pkgconfig
+    - pkg-config
+
+  host:
+    # for library pkg, libmxnet
+    - openblas
+    - llvm-openmp  # [unix]
+    - mkl  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
+    - mkl-include  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
```
Comment (ngam) [inline recipes/mxnet/meta.yaml:89]:
```suggestion
    - mkl  # [target_platform != "osx-arm64"]
    - mkl-include  # [target_platform != "osx-arm64"]
```

File: recipes/mxnet/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    - dmlc  # [unix and (target_platform != "osx-arm64")]
+    # - libtvm  # [unix and x86_64]
+    # - tvm-py  # [unix and x86_64]
+    # - nnvm  # [unix and x86_64]
+
+  run:
+    # for library pkg, libmxnet 
+    - {{ pin_compatible('openblas') }}
+    - {{ pin_compatible('llvm-openmp') }}  # [unix]
+    - {{ pin_compatible('mkl') }}  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
```
Comment (ngam) [inline recipes/mxnet/meta.yaml:113]:
```suggestion
    - {{ pin_compatible('mkl') }}  # [target_platform != "osx-arm64"]
```

**Comment by ngam** [comment]
Status update on this: I don't want to deal with the cuda builds for now; hence, skipping. When someone else has the time and energy to do these runs locally, we can do that.

**Comment by ngam** [comment]
Hi @hmaarrfk, I am skipping the cuda builds. There was a simple error with finding some library that I believe I could resolve; but that's moot because the cuda builds won't finish in under 6 hours. So I think for now I will take this to a feedstock, work on refining the recipe, and then if someone wants to enable the cuda builds, they are welcome (and will likely have to do some local building...)

Are you going to be able to merge, @hmaarrfk? Or should I ping the @conda-forge/staged-recipes again?

File: recipes/mxnet/build.sh [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+else
+    export MXNET_LIBRARY_PATH=${PREFIX}/lib/libmxnet.so
+fi
+
+export MXNET_INCLUDE_PATH=${PREFIX}/include/mxnet
+
+${PYTHON} -m pip install . -vv
+
+#
+# if [[ "$OSTYPE" == "darwin"* ]]; then
```
Comment (hmaarrfk) [inline recipes/mxnet/build.sh:67]: is this still useful?

**Comment by hmaarrfk** [comment]
@conda-forge/staged-recipes this is ready for review

**Comment by ngam** [comment]
cc @jakirkham and @barry-jin

**Comment by ngam** [comment]
Gentle nudge to @conda-forge/staged-recipes and @jakirkham to review this. Also a call for helpful volunteers to maintain this with me

**Comment by ngam** [comment]
[closing due to inactivity.]

**Comment by jakirkham** [comment]
Sorry for the lack of reply. Happy to review, but it may take a bit for me to get to unfortunately

**Comment by hmaarrfk** [comment]
This is a rather large package. 23 days is somewhat normal for a small package. 

Given the size, i would try to be a little more patient.

For example, the pytorch package took more than 1 year and we are still working through some of the challenges.

**Comment by ngam** [comment]
Yo, this is nothing personal --- I just don't want to have this on my list of open issues/PRs given the lack of interest and weird/confusing direction upstream. This is not about being impatient or anything like that; if anyone is interested in picking this up, I would consider helping. I was cleaning up cluttered issues and PRs on my list and I realized it would be best to close this (like many other random issues/PRs)

**Comment by ngam** [comment]
In other words, there are so many other things going on that this is just not near the top. It would be a bit of a stretch for someone like me to maintain mxnet on my own anyway, and I was under the impression some people were interested --- I was mistaken. They've been releasing 2.x RCs for like more than a year upstream, so I have no idea what's going on. It is not clear to me if there is any structure upstream or any direction.

**Comment by ngam** [comment]
@hmaarrfk since you're now part of core, do you want me to reopen this and make it into a feedstock? I am happy to maintain initially (but I won't promise a stellar job, especially when it comes to producing cuda builds if they need more than 6 hours...)

**Comment by ngam** [comment]
I will just reopen it and see if we get any comments. I will also have another look at what's going on upstream to see what's up

**Comment by hmaarrfk** [comment]
I'm going to unsubscribe from this. Ping me when you need an other review.

It seems like a compiler thing at this point:
https://github.com/conda-forge/pytorch-cpu-feedstock/pull/95

https://github.com/conda-forge/ctng-compilers-feedstock/pull/88

**Pull request closed without merging**